### PR TITLE
parse literals more like Postgres

### PR DESCRIPTION
**Pull request by tbg**
- rewrote SQL escaping. It'll output an escaped string when it has to, or when
  a single-quote or backslash appears (the former could be written as '', but
  this is more awkward to implement; the latter needs escaping only in an
  escaped string, which would require an extra pass through the string prior
  to beginning the encoding - not worth it).
- no escaping of "; Postgres doesn't do it
- dissect and update all the test cases affected by this and add some new ones.

found via go-fuzz, ping @dvyukov.

> data0: "INSERT INTO O VALUES ('\0\n\n')"
> data1: "INSERT INTO O VALUES ('\\0\\n\\n')"
> panic: not equal

### CODE REVIEW

File: sql/parser/encode.go [hunk 1]
```diff
@@ -38,16 +38,37 @@ var (
...
+				// Make sure \ turns into \\ if we're in an escaped string.
+				buf = append(buf, '\\')
+			}
 		} else {
 			buf = append(buf, '\\')
 			buf = append(buf, encodedChar)
 		}
 	}
 	buf = append(buf, '\'')
+	if strMod != 0 {
```
Comment (petermattis) [inline sql/parser/encode.go]: Why don't you append the `strMod` byte above where the opening `'` is appended?

File: sql/parser/encode.go [hunk 2]
```diff
@@ -38,16 +38,37 @@ var (
...
+	// Changed if escape sequences necessary. See:
+	// http://www.postgresql.org/docs/9.0/static/sql-syntax-lexical.html
+	var strMod byte
+	for _, ch := range in {
+		if encodedChar := encodeMap[ch]; encodedChar != dontEscape {
+			// If a character which we want to escape is found, update strMod.
+			strMod = 'e' // TODO(tschottdorf) should this be 'E'?
+			break
+		}
+	}
```
Comment (petermattis) [inline sql/parser/encode.go]: These encoding routines were taken from vitess. They are also used in squalor where they were rewritten to append multiple non-encoded bytes at a time. Besides being more efficient, that structure might be easier to use here. Take a look at https://github.com/square/squalor/blob/master/encode.go#L100. The thought is to delay outputting the opening `'` until the first escaped character is seen.

File: sql/parser/scan.go [hunk 3]
```diff
@@ -498,29 +500,39 @@ func (s *scanner) scanString(lval *sqlSymType, ch int, allowEscapes bool) bool {
...
+					lval.str = errUnsupportedEscape
+					return false
 				}
-				// TODO(pmattis): Handle other back-slash escapes? Octal? Hexadecimal?
-				// Unicode?
+
+				// If we end up here, it's a redundant escape - simply drop the
+				// backslash. For example, e'\"' is equivalent to e'"', and
+				// e'\a\b' to e'a\b'.
+				start = s.pos
```
Comment (petermattis) [inline sql/parser/scan.go]: Is there a reason for the loosening of escape processing here? I liked the old code where unknown escapes were errors. Actually, I'm not sure the old code was doing that correctly, but that was the desire.

**Comment by petermattis** [comment]
LGTM

File: sql/parser/encode.go (hunk 2 above)
Comment (tschottdorf) [inline sql/parser/encode.go]: `\` isn't a character that needs escaping, except when in a string that needs escaping. Unless backtracking through the already written buffer is worth it, I think you just need a previous single pass through. Or were you talking about something else?

File: sql/parser/scan.go (hunk 3 above)
Comment (tschottdorf) [inline sql/parser/scan.go]: being more like Postgres. There, anything that's not a known escape sequence escapes to itself (so `\a` becomes `a`): http://sqlfiddle.com/#!15/9eecb7db59d16c80417c72d1e1f4fbf1/2100

File: sql/parser/encode.go (hunk 1 above)
Comment (tschottdorf) [inline sql/parser/encode.go]: done. Remnants from earlier code where I hoped not to have to iterate twice.

File: sql/parser/encode.go (hunk 2 above)
Comment (tschottdorf) [inline sql/parser/encode.go]: I've backed out the part of the change that avoided encoding `'` and `\` where possible from this method and rewrote this method. Looks pretty sleek now and the only thing that has changed is that `'` and `\` lead to an escaped string.

File: sql/parser/scan.go (hunk 3 above)
Comment (tschottdorf) [inline sql/parser/scan.go]: I added a ref to the Postgres doc.

**Pull request merged**